agent/engine/task_manager.go

```diff
@@ -663,16 +663,29 @@ func (mtask *managedTask) startContainerTransitions(transitionFunc containerTran
 			reasons = append(reasons, transition.reason)
 			continue
 		}
+
+		// If the container is already in a transition, skip
+		if transition.actionRequired && !cont.SetAppliedStatus(transition.nextState) {
+			// At least one container is able to be moved forwards, so we're not deadlocked
+			anyCanTransition = true
+			continue
+		}
+
 		// At least one container is able to be moved forwards, so we're not deadlocked
 		anyCanTransition = true
 
 		if !transition.actionRequired {
-			mtask.handleContainerChange(dockerContainerChange{
-				container: cont,
-				event: DockerContainerChangeEvent{
-					Status: transition.nextState,
-				},
-			})
+			// Updating the container status without calling any docker API, send in
+			// a goroutine so that it won't block here before the waitForContainerTransitions
+			// was called after this function.
+			go func(cont *api.Container, status api.ContainerStatus) {
+				mtask.dockerMessages <- dockerContainerChange{
```

Comment:
For my understanding - how do we still handle the container status changes here that `handleContainerChange` does when no action is required for the transition ?

Reply:
All the events sent to `mtask.dockerMessage` will be handled by `handleContainerChange` which works the same as before except that we don't call `handleContainerChange` directly now. Is that what you are asking?

Reply:
>All the events sent to `mtask.dockerMessage` will be handled by `handleContainerChange` which works the same as before except that we don't call `handleContainerChange` directly now

It'd be awesome if you added that as code comment

Reply:
@richardpen Got it, thanks